docs/design/document-type.md

```diff
@@ -14,346 +14,164 @@ such as `struct`, `union`, etc. which are handled by the [Kotlin Smithy SDK](kot
-        require(content[this] == null) {"Key $this is already registered in builder"}
-        content[this] = value.toDocument()
+    class Builder internal constructor() {
+        infix fun kotlin.String.to(value: kotlin.Number)
+        infix fun kotlin.String.to(value: kotlin.Boolean)
+        infix fun kotlin.String.to(value: kotlin.String)
+        infix fun kotlin.String.to(value: Document)
+        infix fun kotlin.String.to(value: Transformable)
     }
 
     /**
-     * Adds given [value] as [SmithyString] to the current [SmithyMap] with [this] as a key
+     * Implemented by structures that can be transformed into a serialized Document.
+     * Must serialize to JSON.
      */
-    infix fun String.to(value: String) {
-        require(content[this] == null) {"Key $this is already registered in builder"}
-        content[this] = value.toDocument()
+    interface Transformable {
+        fun serialize(): kotlin.String
     }
-
-    /**
-     * Adds given [value] to the current [SmithyMap] with [this] as a key
-     */
-    infix fun String.to(value: Document) {
-        require(content[this] == null) {"Key $this is already registered in builder"}
-        content[this] = value
-    }
-}
-
-/**
- * Builds [Document] with given [init] builder.
- *
- * ```
- * val doc = document {
- *     "foo" to 1
- *     "baz" to document {
- *         "quux" to documentArray {
- *             +n(202L)
- *             +n(12)
- *             +true
- *             +"blah"
- *         }
- *     }
- *     "foobar" to document {
- *         "nested" to "a string"
- *         "blerg" to documentArray {
- *             +documentArray {
- *                 +n(2.02)
- *              }
- *         }
- *     }
- * }
- * ```
- *
- * This generates the following JSON:
- * {"foo":1,"baz":{"quux":[202,12,true,"blah"]},"foobar":{"nested":"a string","blerg":[[2.02]]}}
- */
-fun document(init: DocumentBuilder.() -> Unit): Document {
-    val builder = DocumentBuilder()
-    builder.init()
-    return SmithyMap(builder.content)
 }
 ````
 
 Example usage of building a doc or processing one:
 
 ```kotlin
-fun foo() {
-    val doc = document {
+import kotlinx.serialization.json.*
+
+// implement Document.Transformable so we can use this in document builder
+// use kotlinx.serialization
+@Serializable
+data class Clazz(
+    val name: String,
+    val ival: Int
+    ) : Document.Transformable {
+    
+    override fun serialize() = Json.encodeToString(this)
+}
+
+fun main() {
+    val meta = Document {
         "foo" to 1
-        "baz" to document {
-            "quux" to documentArray {
-                +n(202L)
-                +n(12)
-                +true
-                +"blah"
-            }
+        "baz" to Document {
+            "quux" to Document.listOf(
+                202L,
+                12,
+                false,
+                true,
+                "blah",
+                null,
+                Document.listOf(1, null, 2)
+            )
         }
```

Comment:
**Question**: This feels verbose. Does it hamper readability too much to use shorter names like `doc` and `arr`?

```kotlin
"baz" to doc {
    "quux" to arr(
        202L,
        12,
        false,
        true,
        "blah",
        null,
        arr(1, null, 2),
    )
}
```

Reply:
`Document` was to mimic how our codegen request structures work ie.

```kotlin
val request = ListObjectsRequest {
   // ...
   foo = "blah"
}
```

Don't feel strongly about it. kotlinx serialization builder, commented above, doesn't do it that way at all.